rfd/xxxx-config-resource-docs.md

```diff
@@ -0,0 +1,568 @@
+---
+authors: Paul Gottschling (paul.gottschling@goteleport.com)
+state: draft
+title: NNN - Automatically Generate the Configuration Resource Reference
+---
+
+## Required Approvers
+
+- Engineering: @codingllama
+- Product: (@alexfornuto || @xinding33)
+
+## What
+
+Automatically generate the "Dynamic resources" section of the configuration
+resource reference (`docs/pages/reference/resources.mdx`), a list of the
+resources you can apply via `tctl`, from the Teleport source code. 
+
+To do this, write a program that we can run as a new Make target in a clone of
+`gravitational/teleport`. When we make changes to Teleport's dynamic
+configuration resources, we can run the program manually, generate a new
+iteration of the reference, and manually open a pull request with the new
+content.
+
+## Why
+
+We want to ensure that the configuration resource reference is complete and up
+to date so Teleport operators can manage their clusters without creating support
+load.
+
+In general, we should strive to automatically generate reference guides whenever
+possible to ensure accuracy. While adding entries incrementally over time can be
+sustainable at a certain scale, adding many entries at once is a daunting task
+that is difficult for a documentation team to schedule. Our `tctl` resource
+reference is particularly lagging, with only five resources documented at the
+[time of
+writing](https://github.com/gravitational/teleport/blob/58f20f0c3fcc9bb281528915125b95dfaee6cec5/docs/pages/reference/resources.mdx).
+
+## Details
+
+The generator reads a [configuration](#configuration) that specifies struct
+types to generate references from. After parsing Go source files, the generator
+uses the resulting AST information to [populate a template](#the-output). 
+
+To do so, it builds [mappings](#working-with-parsed-go-source) that it can use
+to track type declarations in source packages. Using these mappings, it
+[converts](#converting-structs-to-resources) struct types to template data. To
+document the types of struct fields, the generator writes [additional
+entries](#processing-field-types) to the reference template as appropriate.
+
+### Configuration
+
+The generator will have access to a mapping of package paths to lists of named
+struct types within each package. It will use this mapping to identify the
+struct types to use for generating the resource reference. 
+
+Since struct definitions for `tctl` resources include field types that also
+require their own entries in the resource reference (e.g., `types.Metadata`),
+the configuration only names the _root_ types for the generator to examine.
+
+We don't have a formal naming convention for the root types of each resource,
+and may not want to list *all* resources (e.g., there may be internal resources
+to keep hidden), so it would make sense to list resources explicitly. The CRD
```

Comment:
I suggested manually listing the resources because that's how we do it with the [Kubernetes operator](https://github.com/gravitational/teleport/blob/09adc3ea5cddac5c1fea479ddbfc0075c864677f/integrations/operator/crdgen/main.go#L97-L113). Since we need to develop controllers for the operator, this makes sense in that context, but you're right that it's not great for `tctl`.

I like the idea of including all resources with an embedded `Metadata` field.

Reply:
@codingllama @zmb3 Come to think of it, are there any structs with a `github.com/gravitational/teleport/api/types.Metadata` field that we would want to exclude? Would it be possible to eliminate the configuration entirely?

Reply:
Not sure, but that should be easy to search for. If there aren't any, +1 to automatically finding resources.